File: public/tools/utils/numberpool.ts
import Queue from "queue";

class NumberPool {
	private max: number = 0;
	private allocated: Set<number> = new Set<number>();
	private released: Array<number> = [];

	constructor(private limit: number = 65535) {
	}

	canAllocate(): boolean {
		return this.released.length || this.max < this.limit;
	}

	allocate(): number {
		let n: number;

		if (this.released.length) {
			n = this.released.pop();
		} else if (this.max < this.limit) {
			n = ++this.max;
		} else {
			throw new Error("Unable to allocate next number (limit reached)");
		}

		this.allocated.add(n);
		return n;
	}

	release(n: number): void {
		if (this.allocated.has(n)) {
			this.allocated.delete(n);
			this.released.push(n);
		}
	}
}

export default NumberPool;

File: shared/src/main/scala/gtp3/NumberPool.scala
package gtp3

import scala.collection.mutable

class NumberPool(val limit: Int = 0xFFFF) {
	private var max: Int = 0
	private val allocated = mutable.Set[Int]()
	private val released = mutable.Queue[Int]()

	private def selectNext() = {
		if (released.nonEmpty) {
			released.dequeue()
		} else if (max < limit) {
			max += 1
			max
		} else {
			throw new Exception("Unable to allocate next number (limit reached)")
		}
	}

	def next = {
		val n = selectNext()
		allocated.add(n)
		n
	}

	def release(n: Int) = {
		if (allocated.remove(n)) {
			this.released.enqueue(n)
		}
	}

	def clear() = {
		max = 0
		allocated.clear()
		released.clear()
	}
}
